src/jit/lir.cpp

```diff
@@ -1110,6 +1113,28 @@ bool LIR::Range::CheckLIR(Compiler* compiler, bool checkUnusedValues) const
     // Make sure the range itself is valid.
     assert(IsValid());
 
+    // Make sure the list is well-formed: no node appears more than once (no circularities). Insert each node in a set
+    // and assert we don't find the node again.
+    typedef SimplerHashTable<GenTreePtr, PtrKeyFuncs<GenTree>, bool, JitSimplerHashBehavior> NodeToBoolMap;
+    NodeToBoolMap* nodes = new (compiler->getAllocatorDebugOnly()) NodeToBoolMap(compiler->getAllocatorDebugOnly());
```

Comment:
The problem is, you need to first clear all the mark bits, which requires iterating over the links you're checking for legality. Also, I generally prefer that debug functions don't mutate the things they're checking.

I'll change this to the two pointer, fast/slow algorithm.

Another property I want to check is that the gtPrev list is precisely the opposite of the gtNext list.